### PR TITLE
Freeze top-level val.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt [hunk 1]
```diff
@@ -292,8 +292,11 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
         LLVMBuildStore(builder, value, ptr)
     }
 
-    fun storeAny(value: LLVMValueRef, ptr: LLVMValueRef) {
+    fun storeAny(value: LLVMValueRef, ptr: LLVMValueRef, shallFreeze: Boolean = false) {
         if (isObjectRef(value)) {
+            if (shallFreeze) {
+                call(context.llvm.freezeSubgraph, listOf(value),  Lifetime.IRRELEVANT, ExceptionHandler.Caller)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt]: Proper context-dependent `ExceptionHandler` must be used here instead.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt [hunk 2]
```diff
@@ -292,8 +292,11 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
         LLVMBuildStore(builder, value, ptr)
     }
 
-    fun storeAny(value: LLVMValueRef, ptr: LLVMValueRef) {
+    fun storeAny(value: LLVMValueRef, ptr: LLVMValueRef, shallFreeze: Boolean = false) {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt]: `shallFreeze` seems to be not that closely related to storing a value to memory. It makes some sense to provide `freeze` function here and call it where required.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 3]
```diff
@@ -51,6 +51,9 @@ private val threadLocalAnnotationFqName = FqName("kotlin.native.ThreadLocal")
 val IrClassSymbol.objectIsShared get() =
     !descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName)
 
+val IrFieldSymbol.isShared get() =
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: We should avoid using `Ir*Symbol` classes in signatures and use IR declaration classes instead.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 4]
```diff
@@ -398,7 +404,21 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
+                                }
+                            }
+                    ret(null)
+                }
+
+                appendingTo(bbLocalInit) {
+                    context.llvm.fileInitializers
+                            .forEach {
+                                if (it.initializer?.expression !is IrConst<*>?) {
+                                   if (!it.symbol.isShared) {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: I believe the same check is required when generating `bbLocalDeinit`.

File: runtime/src/main/cpp/Runtime.cpp [hunk 5]
```diff
@@ -78,10 +79,14 @@ RuntimeState* initRuntime() {
...
+  bool firstRuntime = atomicAdd(&aliveRuntimesCount, 1) == 1;
   // Keep global variables in state as well.
-  InitOrDeinitGlobalVariables(INIT_GLOBALS);
-  konan::consoleInit();
-  atomicAdd(&aliveRuntimesCount, 1);
+  if (firstRuntime) {
+    konan::consoleInit();
+    InitOrDeinitGlobalVariables(INIT_GLOBALS);
+  } else {
+    InitOrDeinitGlobalVariables(INIT_THREAD_LOCAL_GLOBALS);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Runtime.cpp]: `INIT_THREAD_LOCAL_GLOBALS` appears to be asymmetric with `DEINIT_THREAD_LOCAL_GLOBALS`:
the former is not called for the first runtime, while the latter is called for the last runtime.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 4 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Yep.

**Pull request merged**